### ISSUE TITLE
Editing model for the memory viewer, feedback needed

### ISSUE THREAD

**Issue by haneefdm** [haneefdm/memview#3]
This is a tricky one for me as we are not talking about editing files here. I only envision limited editing -- a byte here, a byte there, kind of thing. I am not expecting major bulky edits. Regardless, here are a couple of models

First of all, edits will only be allowed when the debugger is in paused state. Keep in mind that we are not part of the debugger or VSCode debug infrastructure. We are observers. We only find out the current state after a transition has already occurred. For instance, when a `continue` occurs, it takes us a while for us to find out.

## Saving model (this must be decided first)
1. Save as you go. As you modify each byte, it will be saved instantly if possible. If the write fails, the user is notified and we revert the byte(s) back to its original state. Writes can fail because perhaps the debugger already continued or perhaps we are looking at a non-writable section. Even though the debugger supports write requests, you could still be looking at areas that are not writable. You will have a visual indication as to which bytes have been modified.
2. Batch all the edits until you hit the `Save` button (explicit save command). This is what normal editors do. The danger here is that the user may have continued while some edits are unsaved. We will lose those edits.

Would be good to find out what other IDEs do.

## Undo/Redo
1. All undo/redo stacks are on a per stopped-session basis. ie, when the program continues, the undo stack will be cleared. Will be enabled on the next stop.
2. Undo/Redo will follow whatever saving model we use above

## Cut/Copy/Paste
1. `Cut`: not allowed. Will be the same as Copy. This means, that we will not shift all the memory left when a byte is cut. Remember we are kind of an infinite scroll....so.
2. `Copy`: We put the selected cell(s) into the clipboard. But, what to put in the clipboard.. Surely not the hex bytes? Or the hex-bites as one format (text) and binary data as well. There may be some work to be done for how to select multiple disjoint cells, but lower priority.
4. `Paste`: Not sure how the clipboard works. It can contain multiple random things. Sure we can accept anything because it is random binary data. But doesn't seem right. It won't even work

It has been a long time since I worked with Clipboards so, I have to take inspiration from others. HexEdit is only place to look and maybe even make the formats compatible.

**Comment by benmcmorran** [comment]
One thought that comes to mind is that other extensions will likely already know which memory sections are writable (e.g. from an SVD). I haven't thought through the details too much yet, but it could make sense to provide an extension API so that the memory viewer knows which sections are safe to write to (or even read from). We end up doing something like this in our [peripheral view](https://marketplace.visualstudio.com/items?itemName=ms-vscode.vscode-embedded-tools), where we attempt to read memory in large chunks but break up the reads as needed when there are unreadable sections specified in the SVD.

**Comment by haneefdm** [comment]
This issue is now at https://github.com/mcu-debug/memview/issues/4
